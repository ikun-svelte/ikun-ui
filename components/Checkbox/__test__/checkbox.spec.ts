import { tick } from 'svelte';
import { afterEach, expect, test, vi, describe, beforeEach } from 'vitest';
import KCheckBox from '../src/index.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KCheckBox', () => {
	test('props: value', async () => {
		let value = true;
		const instance = new KCheckBox({
			target: host,
			props: {
				value
			}
		});
		instance.$on('updateValue', (v) => {
			value = v.detail;
		});
		const input = host.getElementsByTagName('input')[0];
		expect(instance).toBeTruthy();
		expect(input.value === 'true').toBeTruthy();
		input.click();
		await tick();
		instance.$set({ value });
		await tick();
		expect(value).not.toBeTruthy();
		expect(input.value === 'false').toBeTruthy();

		initHost();

		let value2 = false;
		const instance2 = new KCheckBox({
			target: host,
			props: {
				value: value2
			}
		});
		instance2.$on('updateValue', (v) => {
			value2 = v.detail;
		});
		const input2 = host.getElementsByTagName('input')[0];
		expect(instance2).toBeTruthy();
		expect(input2.value === 'false').toBeTruthy();
		input2.click();
		await tick();
		instance2.$set({ value: value2 });
		await tick();
		expect(value2).toBeTruthy();
		expect(input2.value === 'true').toBeTruthy();
	});

	test('props: label', async () => {
		const instance = new KCheckBox({
			target: host,
			props: {
				label: 'foo'
			}
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('foo')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: size', async () => {
		const instance = new KCheckBox({
			target: host,
			props: {
				size: 'lg'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-checkbox--lg')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
		instance.$set({ size: 'md' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-checkbox--md')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
		instance.$set({ size: 'sm' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-checkbox--sm')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		const instance = new KCheckBox({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-cur-disabled')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-checkbox--box__disabled')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: indeterminate', async () => {
		const instance = new KCheckBox({
			target: host,
			props: {
				indeterminate: true
			}
		});
		await tick();
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('i-carbon-subtract')).toBeTruthy();
		instance.$set({ indeterminate: false });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('i-carbon-subtract')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: checkColor & bgCheckColor ', async () => {
		new KCheckBox({
			target: host,
			props: {
				value: true,
				checkColor: 'text-red',
				bgCheckColor: 'blue'
			}
		});
		await tick();
		expect(host.innerHTML.includes('border-ikun-main" style="background-color: blue')).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-checkmark text-red')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: bgUnCheckColor ', async () => {
		new KCheckBox({
			target: host,
			props: {
				value: false,
				bgUnCheckColor: 'green'
			}
		});
		await tick();
		expect(
			host.innerHTML.includes('k-checkbox--box" style="background-color: green;')
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: should trigger updateValue event', async () => {
		const mockFn = vi.fn();
		let value = false;
		const instance = new KCheckBox({
			target: host,
			props: {
				value
			}
		});
		await tick();
		instance.$on('updateValue', (v) => {
			value = v.detail;
			mockFn();
		});
		const input = host.getElementsByTagName('input')[0];
		input.click(); // or input.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(instance).toBeTruthy();
		expect(value).toBeTruthy();
		expect(mockFn).toBeCalled();
	});

	test('event: should not trigger updateValue event when disabled', async () => {
		const mockFn = vi.fn();
		let value = false;
		const instance = new KCheckBox({
			target: host,
			props: {
				disabled: true,
				value
			}
		});
		await tick();
		instance.$on('updateValue', (v) => {
			value = v.detail;
			mockFn();
		});
		const input = host.getElementsByTagName('input')[0];
		input.click(); // or input.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(instance).toBeTruthy();
		expect(value).not.toBeTruthy();
		expect(mockFn).not.toBeCalled();
	});
});
