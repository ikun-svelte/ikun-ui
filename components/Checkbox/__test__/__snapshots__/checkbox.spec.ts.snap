// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KCheckBox > props: bgUnCheckColor  1`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box" style="background-color: green;"> </div> </label>"`;

exports[`Test: KCheckBox > props: checkColor & bgCheckColor  1`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="true" hidden=""> <div class="k-checkbox--box bg-ikun-main border-ikun-main" style="background-color: blue;"><div><span class="k-icon--base k-icon-transition i-carbon-checkmark text-red" role="" aria-hidden="true"></span></div> </div> </label>"`;

exports[`Test: KCheckBox > props: disabled 1`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="false" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled"> </div> </label>"`;

exports[`Test: KCheckBox > props: indeterminate 1`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box"> </div> </label>"`;

exports[`Test: KCheckBox > props: label 1`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box"> </div> <span class="k-checkbox--label">foo</span></label>"`;

exports[`Test: KCheckBox > props: size 1`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--lg"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box"> </div> </label>"`;

exports[`Test: KCheckBox > props: size 2`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box"> </div> </label>"`;

exports[`Test: KCheckBox > props: size 3`] = `"<label class="k-checkbox--base k-checkbox--base__dark k-checkbox--sm"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box"> </div> </label>"`;
